### PR TITLE
Survival analysis balancing weights QP v1

### PR DESCRIPTION
**Pull request by shenmaya**
Code for finding survival analysis balancing weights via QP. 

Version 1: no delta, separately optimize weights for each time t

### CODE REVIEW

File: R/survival_qp.R [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+#' @return \itemize{
+#'          \item{weights}{Estimated weights as a length n vector}
+#'          \item{imbalance}{Imbalance in covariates as a d X J matrix}}
+#' @export
+survival_qp <- function(B_X, trt, times, events, t, lambda = 0, lowlim = 1, uplim = NULL,
+                        verbose = TRUE, eps_abs = 1e-5, eps_rel = 1e-5, ...) {
+  # Convert X to a matrix
+  B_X <- as.matrix(B_X)
+  
+  # Split data and treatment by factor
```
Comment (ebenmichael) [inline R/survival_qp.R]: What is this splitting the data on?

File: R/survival_qp.R [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    stop("Lower threshold must be lower than upper threshold")
+  }
+  if(lowlim < 0) {
+    stop("Lower threshold must be at least 0")
+  }
+  if(uplim > n) {
+    stop("Upper threshold must be at most n (",
+         n, ")")
+  }
+}
```
Comment (ebenmichael) [inline R/survival_qp.R]: Are these necessary? We might want to consider having negative weights at some point maybe.

File: R/survival_qp.R [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+#' @param t Time
+#' @param lowlim Lower limit on weights, default 1
+#' @param uplim Upper limit on weights, default NULL
+#' @param verbose Whether to show messages, default T
+#' 
+#' @return A, l, and u
+create_constraints <- function(n, trt, times, events, t, lowlim = 1, uplim = NULL, verbose = TRUE) {
+
+  # Vector of booleans indicating if indiv is not censored at time t
+  noncens_t <- ((times >= t) & (events == TRUE)) | (events == FALSE)
```
Comment (ebenmichael) [inline R/survival_qp.R:135]: So I understand: `(times >= t)` means survived or not censored by at least `t` and `(times >= t) & (events == TRUE)` means individuals that are censored  but at time `t` or later?  Then `((times >= t) & (events == TRUE)) | (events == FALSE)` are individuals that are censored at time `t` or later or fail before they are censored?

File: R/survival_qp.R [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+  if(verbose) message("\tx Summed weight constraint for non-censored treatment group")
+  A1 <- as.integer(noncens_t & (trt == 1))
+  l1 <- n
+  u1 <- n
+  
+  if(verbose) message("\tx Summed weight constraint for non-censored control group")
+  A2 <- as.integer(noncens_t & (trt == 0))
+  l2 <- n
+  u2 <- n
```
Comment (ebenmichael) [inline R/survival_qp.R:145]: I think these should solve separate optimization problems for the treated and control units. You can write one version of the code e.g. for only treated units, then pass in `1-trt` to have it work for the control units. Right now I think this mixes the two and doesn't get balance on either.

File: R/survival_qp.R [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+  solution <- osqp::solve_osqp(P, q, constraints$A,
+                               constraints$l, constraints$u,
+                               settings)
+  
+  weights <- solution$x
+  
+  # imbalance_root <- t(w) %*% B - Bbar_X
+  # imbalance <- imbalance_root %*% t(imbalance_root)
+  
+  imbalance <- sum(((t(weights) %*% B_X) - Bbar_X)^2)
```
Comment (ebenmichael) [inline R/survival_qp.R]: maybe change this to mean so that it's the mean square imbalance

**Review by ebenmichael — CHANGES_REQUESTED** [review]
The biggest thing: I think that the optimization problem should be separate for treated and control. This version of the code mixes the objectives so I don't think the weights are actually targeting for balance in both uncensored treated and control to the full population, but rather just for uncensored to the full population.

File: R/survival_qp.R (hunk 3 above)
Comment (shenmaya) [reply R/survival_qp.R:135]: `noncens_t` indicates whether an individual was either 1. an individual that was not censored at time t (but was censored later) - `(times >= t) & (events == TRUE)`, or 2. an individual that is never censored - `events == FALSE`. I was under the impression that we wanted weights for both of these groups, not just individuals who survived past time t - `(times >= t)`?

File: R/survival_qp.R [hunk 6]
```diff
@@ -66,8 +66,8 @@ survival_qp <- function(B_X, trt, times, events, t, lambda = 0, lowlim = 1, upli
   # imbalance_root <- t(w) %*% B - Bbar_X
   # imbalance <- imbalance_root %*% t(imbalance_root)
   
-  imbalance <- sum(((t(weights) %*% B_X) - Bbar_X)^2) 
-    
+  imbalance <- mean(((t(weights) %*% B_X) - Bbar_X)^2) 
+
```
Comment (ebenmichael) [inline R/survival_qp.R]: Can this be separated out into the vector of imbalances for treated and control units? Then we can check whether this is working by checking whether the imbalance is small.

File: R/survival_qp.R [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+  
+  weights <- solution$x
+
+  # Return vectors of imbalances for treated and control 
+  noncens_t <- ((times >= t) & (events == TRUE)) | (events == FALSE)
+  B0_X <- B_X*(trt == 0)
+  B1_X <- B_X*(trt == 1)
+  
+  imbalance0 <- (t(B0_X) %*% Matrix(weights*noncens_t, n, 1)) - Matrix(Bbar_X, k, 1)
+  imbalance1 <- (t(B1_X) %*% Matrix(weights*noncens_t, n, 1)) - Matrix(Bbar_X, k, 1)
```
Comment (ebenmichael) [inline R/survival_qp.R]: These imbalance measures are unnormalized, it should be, e.g.
```
  imbalance0 <- colSums(B_X * (trt == 0) * weights * noncens_t) / n - Bbar_X
  imbalance1 <- colSums(B_X * (trt == 1)* weights * noncens_t) / n - Bbar_X
```

File: R/survival_qp.R [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+#' Create the q vector for an QP that solves min_x 0.5 * x'Px + q'x
+#' @param B_X n x k basis matrix of covariates 
+#' @param Bbar_X Vector of population means to re-weight to
+#'
+#' @return q vector
+create_q_vector_surv <- function(n, trt, B_X, Bbar_X) {
+  B0_X <- B_X*(trt == 0)
+  B1_X <- B_X*(trt == 1)
+
+  q <- -(2/n) * Bbar_X %*% t(B0_X + B1_X)
```
Comment (ebenmichael) [inline R/survival_qp.R]: `osqp` is assuming that you're passing in an optimization problem as $\frac{1}{2}x'Px + q'x$, so this shouldn't have the multiplication by 2, i.e.
```
q <- -(1/n) * Bbar_X %*% t(B0_X + B1_X)
```

**Review by ebenmichael — CHANGES_REQUESTED** [review]

**Pull request merged**